commit ade77f3c1499e591ba6ac38449c437f1d80af17a
Author: Robert Adam <dev@robert-adam.de>
Date:   Mon Dec 6 11:19:22 2021 +0100

    BUILD: Fix violation of ODR
    
    The "ODR" (One Definition Rule) states that any function/object/etc. may
    only be defined once. However, since we are using some preprocessor
    magic to only include certain fields in the Channel struct, we have to
    make sure that everywhere Channel.h is included, these macros are set
    accordingly.
    
    However, the User.h was part of the "shared" library and in that case no
    macros were being set when compiling this translation unit. In the
    actual client however, the "MUMBLE" macro is set, such that all headers
    being part of the "mumble" target will use a different Channel
    definition than the User class which is part of the "shared" target.
    
    In order to not violate the ODR, the User class was not excluded from
    being part of the "shared" target and is instead explicitly included in
    the "mumble" and "mumble-server" targets (making sure the necessary
    macros will be set).

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index bfff395e9..2a6c58aa1 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -72,7 +72,6 @@ set(SHARED_SOURCES
 	"SSLLocks.cpp"
 	"Timer.cpp"
 	"UnresolvedServerAddress.cpp"
-	"User.cpp"
 	"Version.cpp"
 
 	"crypto/CryptographicHash.cpp"
@@ -108,7 +107,6 @@ set(SHARED_HEADERS
 	"SSLLocks.h"
 	"Timer.h"
 	"UnresolvedServerAddress.h"
-	"User.h"
 	"Version.h"
 
 	"crypto/CryptographicHash.h"
diff --git a/src/mumble/CMakeLists.txt b/src/mumble/CMakeLists.txt
index bda88d0c4..0d53735af 100644
--- a/src/mumble/CMakeLists.txt
+++ b/src/mumble/CMakeLists.txt
@@ -297,6 +297,8 @@ set(MUMBLE_SOURCES
 	"${SHARED_SOURCE_DIR}/Group.cpp"
 	"${SHARED_SOURCE_DIR}/Group.h"
 	"${SHARED_SOURCE_DIR}/SignalCurry.h"
+	"${SHARED_SOURCE_DIR}/User.cpp"
+	"${SHARED_SOURCE_DIR}/User.h"
 
 	"${3RDPARTY_DIR}/smallft/smallft.cpp"
 
diff --git a/src/murmur/CMakeLists.txt b/src/murmur/CMakeLists.txt
index 112c919c0..cc3860262 100644
--- a/src/murmur/CMakeLists.txt
+++ b/src/murmur/CMakeLists.txt
@@ -47,6 +47,8 @@ set(MURMUR_SOURCES
 	"${SHARED_SOURCE_DIR}/Connection.h"
 	"${SHARED_SOURCE_DIR}/Group.cpp"
 	"${SHARED_SOURCE_DIR}/Group.h"
+	"${SHARED_SOURCE_DIR}/User.cpp"
+	"${SHARED_SOURCE_DIR}/User.h"
 )
 
 if(WIN32)
